gemd/util/impl.py

```diff
@@ -98,34 +109,118 @@ def _substitute(thing: Any,
         visited = {}
     if thing.__hash__ is not None and thing in visited:
         return visited[thing]
+
     if applies(thing):
         replacement = sub(thing)
         if thing.__hash__ is not None:
             visited[thing] = replacement
-        new = _substitute(replacement, sub, applies, visited)
-    elif isinstance_list(thing):
-        new = [_substitute(x, sub, applies, visited) for x in thing]
-    elif isinstance_tuple(thing):
-        new = tuple(_substitute(x, sub, applies, visited) for x in thing)
-    elif isinstance_dict(thing):
+    else:
+        replacement = thing
+
+    if _cached_isinstance(replacement, list):
+        new = [_substitute(x, sub, applies, visited) for x in replacement]
+    elif _cached_isinstance(replacement, tuple):
+        new = tuple(_substitute(x, sub, applies, visited) for x in replacement)
+    elif _cached_isinstance(replacement, dict):
         new = {_substitute(k, sub, applies, visited): _substitute(v, sub, applies, visited)
-               for k, v in thing.items()}
-    elif isinstance_dict_serializable(thing):
+               for k, v in replacement.items()}
+    elif _cached_isinstance(replacement, DictSerializable):
         new_attrs = {_substitute(k, sub, applies, visited): _substitute(v, sub, applies, visited)
-                     for k, v in thing.as_dict().items()}
-        new = thing.build(new_attrs)
+                     for k, v in replacement.as_dict().items()}
+        new = replacement.build(new_attrs)
     else:
-        new = thing
+        new = replacement
 
     if thing.__hash__ is not None:
         visited[thing] = new
-    if new.__hash__ is not None:
-        visited[new] = new
 
-    # assert type(thing) == type(new), "{} is not {}".format(type(thing), type(new))
     return new
 
 
+def _substitute_inplace(thing: Any,
+                        sub: Callable[[object], object],
+                        applies: Callable[[object], bool],
+                        visited: Dict[object, object] = None) -> object:
```

Comment:
I guess this should be `visited: Optional[Dict[object, object]]`

Reply:
It's a weird case because the user should never provide the `visited` argument, and None is an internal signifier that this is the first frame.